File: R/R/annotate.R
#' Annotate media bias in a headline edit
#'
#' Sends an original and edited headline to the OpenAI API and returns a
#' structured analysis of 13 types of media bias.
#'
#' @param original Character string. The original headline.
#' @param edited Character string. The edited headline.
#' @param added_words Character vector of words added in the edit, or NULL
#'   to compute automatically via word-level set difference.
#' @param removed_words Character vector of words removed in the edit, or NULL
#'   to compute automatically via word-level set difference.
#' @param api_key Character string. OpenAI API key. Defaults to the
#'   \code{OPENAI_API_KEY} environment variable.
#' @param model Character string. The OpenAI model to use. Defaults to
#'   \code{"gpt-3.5-turbo"}.
#' @return A named list with each bias type containing \code{value}
#'   ("Added"/"Removed"/"None") and \code{reason}.
#' @export
annotate_bias <- function(original,
                          edited,
                          added_words = NULL,
                          removed_words = NULL,
                          api_key = Sys.getenv("OPENAI_API_KEY"),
                          model = "gpt-3.5-turbo") {

  if (is.null(api_key) || nchar(api_key) == 0) {
    stop("OpenAI API key not found. Set the OPENAI_API_KEY environment variable or pass it via the api_key argument.")
  }

  # Compute word-level differences if not provided
  original_words <- strsplit(original, "\\s+")[[1]]
  edited_words <- strsplit(edited, "\\s+")[[1]]

  if (is.null(added_words)) {
    added_words <- setdiff(edited_words, original_words)
  }
  if (is.null(removed_words)) {
    removed_words <- setdiff(original_words, edited_words)
  }

  added_str <- if (length(added_words) == 0) "None" else paste(added_words, collapse = ", ")
  removed_str <- if (length(removed_words) == 0) "None" else paste(removed_words, collapse = ", ")

  # Build the user message

  user_message <- paste0(
    "Original Headline: ", original, "\n",
    "Edited Headline: ", edited, "\n",
    "Added words: ", added_str, "\n",
    "Removed words: ", removed_str
  )

  # Call the OpenAI Chat Completions API
  body <- list(
    model = model,
    messages = list(
      list(role = "system", content = mediaspin_prompt()),
      list(role = "user", content = user_message)
    ),
    temperature = 0
  )

  resp <- httr2::request("https://api.openai.com/v1/chat/completions") |>
    httr2::req_headers(
      Authorization = paste("Bearer", api_key),
      `Content-Type` = "application/json"
    ) |>
    httr2::req_body_json(body) |>
    httr2::req_perform()

  resp_body <- httr2::resp_body_json(resp)

  # Extract the assistant's reply

  raw_text <- resp_body$choices[[1]]$message$content

  # Parse and return structured result
  parse_bias_response(raw_text)
}


#' Batch annotate media bias for multiple headline pairs
#'
#' Processes a data.frame of original and edited headline pairs, calling
#' \code{annotate_bias} for each row.
#'
#' @param pairs_df A data.frame with at least two columns: \code{original} and
#'   \code{edited}. May optionally include \code{added_words} and
#'   \code{removed_words} columns (character strings with comma-separated words).
#' @param api_key Character string. OpenAI API key. Defaults to the
#'   \code{OPENAI_API_KEY} environment variable.
#' @param model Character string. The OpenAI model to use. Defaults to
#'   \code{"gpt-3.5-turbo"}.
#' @return A list of named lists, one per row, each containing the bias
#'   analysis results.
#' @export
annotate_batch <- function(pairs_df,
                           api_key = Sys.getenv("OPENAI_API_KEY"),
                           model = "gpt-3.5-turbo") {

  if (!all(c("original", "edited") %in% names(pairs_df))) {
    stop("pairs_df must contain 'original' and 'edited' columns.")
  }

  results <- vector("list", nrow(pairs_df))

  for (i in seq_len(nrow(pairs_df))) {
    added <- NULL
    removed <- NULL

    if ("added_words" %in% names(pairs_df) && !is.na(pairs_df$added_words[i])) {
      added <- trimws(strsplit(pairs_df$added_words[i], ",")[[1]])
    }
    if ("removed_words" %in% names(pairs_df) && !is.na(pairs_df$removed_words[i])) {
      removed <- trimws(strsplit(pairs_df$removed_words[i], ",")[[1]])
    }

    message(sprintf("Annotating row %d of %d...", i, nrow(pairs_df)))

    results[[i]] <- tryCatch(
      annotate_bias(
        original = pairs_df$original[i],
        edited = pairs_df$edited[i],
        added_words = added,
        removed_words = removed,
        api_key = api_key,
        model = model
      ),
      error = function(e) {
        warning(sprintf("Error on row %d: %s", i, conditionMessage(e)))
        NULL
      }
    )
  }

  results
}

File: python/mediaspin/annotate.py
"""Core annotation logic for the MediaSpin bias detection pipeline."""

from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Tuple

from openai import OpenAI

from .prompt import BIAS_TYPES, SYSTEM_PROMPT


def _diff_words(original: str, edited: str) -> Tuple[List[str], List[str]]:
    """Compute added and removed words via simple word-level set difference."""
    orig_words = original.split()
    edit_words = edited.split()

    # Use multisets (count occurrences) for accurate diffing
    from collections import Counter

    orig_counts = Counter(orig_words)
    edit_counts = Counter(edit_words)

    added = []
    for word, count in edit_counts.items():
        diff = count - orig_counts.get(word, 0)
        added.extend([word] * max(diff, 0))

    removed = []
    for word, count in orig_counts.items():
        diff = count - edit_counts.get(word, 0)
        removed.extend([word] * max(diff, 0))

    return added, removed


def _format_user_message(
    original: str,
    edited: str,
    added_words: List[str],
    removed_words: List[str],
) -> str:
    """Format the user message for the OpenAI API call."""
    added_str = ", ".join(added_words) if added_words else "(none)"
    removed_str = ", ".join(removed_words) if removed_words else "(none)"
    return (
        f"Original Headline: {original}\n"
        f"Edited Headline: {edited}\n"
        f"Added words: {added_str}\n"
        f"Removed words: {removed_str}"
    )


def parse_response(text: str) -> Dict[str, Any]:
    """Parse the raw LLM response into a structured dictionary.

    Returns a dict with:
        - "words_added": list of {"word": str, "pos": str}
        - "words_removed": list of {"word": str, "pos": str}
        - "bias_analysis": dict mapping each bias type to
          {"value": "Added"/"Removed"/"None", "reason": str}
        - "raw": the original response text
    """
    result: Dict[str, Any] = {
        "words_added": [],
        "words_removed": [],
        "bias_analysis": {},
        "raw": text,
    }

    # Parse Words Added
    added_match = re.search(r"Words Added:\s*(.+)", text)
    if added_match:
        for m in re.finditer(r"(\S+)\s*\[([^\]]+)\]", added_match.group(1)):
            result["words_added"].append({"word": m.group(1).rstrip(","), "pos": m.group(2)})

    # Parse Words Removed
    removed_match = re.search(r"Words Removed:\s*(.+)", text)
    if removed_match:
        for m in re.finditer(r"(\S+)\s*\[([^\]]+)\]", removed_match.group(1)):
            result["words_removed"].append({"word": m.group(1).rstrip(","), "pos": m.group(2)})

    # Parse Bias Analysis list
    # Match lines like: 1. Spin [None]: explanation text
    bias_pattern = re.compile(
        r"^\d+\.\s*(.+?)\s*\[(Added|Removed|None)\]\s*:\s*(.+)",
        re.MULTILINE,
    )
    for m in bias_pattern.finditer(text):
        bias_name = m.group(1).strip()
        value = m.group(2).strip()
        reason = m.group(3).strip()
        result["bias_analysis"][bias_name] = {"value": value, "reason": reason}

    return result


def annotate_bias(
    original: str,
    edited: str,
    added_words: Optional[List[str]] = None,
    removed_words: Optional[List[str]] = None,
    api_key: Optional[str] = None,
    model: str = "gpt-3.5-turbo",
) -> Dict[str, Any]:
    """Annotate bias in a news headline edit.

    Parameters
    ----------
    original : str
        The original headline.
    edited : str
        The edited headline.
    added_words : list of str, optional
        Words added in the edit. Computed automatically if not provided.
    removed_words : list of str, optional
        Words removed in the edit. Computed automatically if not provided.
    api_key : str, optional
        OpenAI API key. If not provided, the OPENAI_API_KEY environment
        variable is used.
    model : str
        OpenAI model to use (default: ``"gpt-3.5-turbo"``).

    Returns
    -------
    dict
        Structured annotation result with keys ``words_added``,
        ``words_removed``, ``bias_analysis``, and ``raw``.
    """
    if added_words is None or removed_words is None:
        auto_added, auto_removed = _diff_words(original, edited)
        if added_words is None:
            added_words = auto_added
        if removed_words is None:
            removed_words = auto_removed

    user_message = _format_user_message(original, edited, added_words, removed_words)

    client = OpenAI(api_key=api_key)
    response = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": user_message},
        ],
        temperature=0.0,
    )

    raw_text = response.choices[0].message.content or ""
    return parse_response(raw_text)


def annotate_batch(
    pairs: List[Tuple[str, str]],
    api_key: Optional[str] = None,
    model: str = "gpt-3.5-turbo",
) -> List[Dict[str, Any]]:
    """Annotate bias for a batch of headline pairs.

    Parameters
    ----------
    pairs : list of (str, str)
        List of ``(original_headline, edited_headline)`` tuples.
    api_key : str, optional
        OpenAI API key.
    model : str
        OpenAI model to use.

    Returns
    -------
    list of dict
        List of structured annotation results.
    """
    results = []
    for original, edited in pairs:
        result = annotate_bias(original, edited, api_key=api_key, model=model)
        results.append(result)
    return results
